src/ds/test/thread_messaging.cpp

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#include "../thread_messaging.h"
+
+#include <doctest/doctest.h>
+
+struct Nothing
+{};
+
+static bool happened = false;
+
+static void always(std::unique_ptr<threading::Tmsg<Nothing>> msg)
+{
+  happened = true;
+}
+
+static void never(std::unique_ptr<threading::Tmsg<Nothing>> msg)
+{
+  CHECK(false);
+}
+
+// Note: this only works with ASAN turned on, which catches m2 not being
+// freed.
```

Comment:
To test this without ASAN, I think you can create a struct that changes some static state in its destructor, then you can confirm that its destructor is/isn't called? Something like:

```
struct Foo {
  static bool destructed = false;
  ~Foo() { CHECK(!destructed); destructed = true; }
};

{
  threading::ThreadMessaging tm;
  auto m1 = std::make_unique<threading::Tmsg<Foo>>(&always);
  CHECK(!Foo::destructed);
  tm.run_one(task);
  // Running a task destroys the attached payload
  CHECK(Foo::destructed);
}

Foo::destructed = false; // RESET

{
   threading::ThreadMessaging tm;
   // add task but don't run
   CHECK(!Foo::destructed);
}

// Destroying ThreadMessaging destroys contained messages
CHECK(Foo::destructed);
```

Reply:
Yeah that checks that Payload is destructed, but it isn't quite the same as checking TMsg is. Still good though.